src/org/opensolaris/opengrok/history/GitRepository.java

```diff
@@ -252,11 +246,154 @@ public InputStream getHistoryGet(String parent, String basename, String rev) {
+        // moved/{renamed.c => renamed2.c} (100%)
+        Pattern pattern = Pattern.compile("\\s*([^\\{]+)\\s*\\{\\s*(.+?)\\s+=>\\s+([^\\}]+)\\s*\\}(.*)");
+        Matcher m = pattern.matcher(renamedFile);
+        if (m.find()) {
+            renamedFile = m.replaceAll("$1$2 => $1$3$4");
+        }
+        return renamedFile;
+    }
+
+    /**
+     * Get the name of file in given revision.
+     *
+     * @param fullpath file path
+     * @param changeset changeset
+     * @return original filename
+     * @throws java.io.IOException
+     */
+    protected String findOriginalName(String fullpath, String changeset)
+            throws IOException {
+        if (fullpath == null || fullpath.isEmpty() || fullpath.length() < directoryName.length()) {
+            LOGGER.log(Level.SEVERE,
+                    "Invalid file path string: {0}", fullpath);
+            return null;
+        }
+
+        if (changeset == null || changeset.isEmpty()) {
+            LOGGER.log(Level.SEVERE,
+                    "Invalid changeset string: {0}", changeset);
+            return null;
+        }
+
+        String file = fullpath.replace(directoryName + File.separator, "");
+        /*
+         * Get the list of file renames for given file to the specified
+         * revision.
+         */
+        String[] argv = {
+            ensureCommand(CMD_PROPERTY_KEY, CMD_FALLBACK),
+            "log",
+            "--follow",
+            "--summary",
+            ABBREV_LOG,
+            "--abbrev-commit",
+            fullpath
+        };
+
+        ProcessBuilder pb = new ProcessBuilder(argv);
+        Process process = Runtime.getRuntime().exec(argv, null, new File(directoryName));
+
+        try {
+            try (BufferedReader in = new BufferedReader(
+                    new InputStreamReader(process.getInputStream()))) {
+                String line;
+                String rev = null;
+                while ((line = in.readLine()) != null) {
+                    if (line.startsWith("commit ")) {
+                        rev = line.replace("commit ", "");
+                        continue;
+                    }
+
+                    if (rev.equals(changeset)) {
+                        break;
+                    }
+
+                    if (line.startsWith(" rename ")) {
+                        line = line.replace(" rename ", "");
+                        line = expandGitRenamedOutput(line);
+                        String[] split = line.split(" => ");
+                        if (split.length < 2) {
+                            throw new IOException(
+                                    String.format(
+                                            "Invalid format of the git output. Expected "
+                                            + "\"file => otherfile\" but got \"{}\"", line));
+                        }
+                        file = split[0];
+                    }
+
+                    if (rev.equals(changeset)) {
```

Comment:
not sure how this is supposed work given the same piece of code on lines 360-362